libfwupd/fwupd-client.c

```diff
@@ -4240,6 +4250,54 @@ fwupd_client_upload_bytes_finish (FwupdClient *self, GAsyncResult *res, GError *
 	return g_task_propagate_pointer (G_TASK(res), error);
 }
 
+/* this is bad; we dlopen libsoup-2.4.so.1 and get the gtype manually
+ * to avoid deps on both libcurl and libsoup whilst preserving ABI */
+static void
+fwupd_client_ensure_soup_session (FwupdClient *self)
+{
+	FwupdClientObjectNewFunc func = NULL;
+	FwupdClientPrivate *priv = GET_PRIVATE (self);
+	GType soup_gtype;
+
+	/* already set up */
+	if (priv->soup_session != NULL)
+		return;
+
+	/* known GType, just create */
+	soup_gtype = g_type_from_name ("SoupSession");
+	if (soup_gtype != 0) {
+		priv->soup_session = g_object_new (soup_gtype, NULL);
+		return;
+	}
+
+	/* load the library at rumtime, leaking the module */
+	if (priv->soup_module == NULL) {
+		g_autofree gchar *fn = NULL;
+		const gchar *libdirs[] = {
+			FWUPD_LIBDIR,
+			"/usr/lib64",
+			"/usr/lib",
+			NULL };
+		for (guint i = 0; libdirs[i] != NULL && priv->soup_module == NULL; i++) {
+			fn = g_build_filename (libdirs[i], "libsoup-2.4.so.1", NULL);
+			g_debug ("looking for libsoup in %s", fn);
+			priv->soup_module = g_module_open (fn, G_MODULE_BIND_LAZY);
+		}
+	}
+	if (priv->soup_module == NULL) {
+		g_warning ("failed to find libsoup library");
+		return;
+	}
+	if (!g_module_symbol (priv->soup_module,
+			      "soup_session_new",
+			      (gpointer *) &func)) {
+		g_warning ("failed to find soup_session_get_type()");
```

Comment:
maybe unload the module in this error path?

Reply:
I thought I'd leave it be, otherwise we'll keep trying to open libsoup for every property read.

Reply:
But don't you want to do to that?  What if someone didn't have libsoup installed and then installed it later at runtime?

Reply:
Possible indeed. Fixed.